src/util/NuulsUploader.cpp

```diff
@@ -0,0 +1,222 @@
+
+        .multiPart(payload)
+        .onSuccess([&textEdit, channel](NetworkResult result) -> Outcome {
+            textEdit.insertPlainText(result.getData() + QString(" "));
+            isUploading = false;
+            if (uploadQueue.empty())
+            {
+                channel->addMessage(makeSystemMessage(
+                    QString("Your image has been uploaded.")));
+            }
+            else
+            {
+                channel->addMessage(makeSystemMessage(
+                    QString("Your image has been uploaded. %1 left. Please "
+                            "wait until all of them are uploaded. About %2 "
+                            "seconds left.")
+                        .arg(uploadQueue.size())
+                        .arg(uploadQueue.size() * (UPLOAD_DELAY / 1000 + 1))));
+                // 2 seconds for the timer that's there not to spam the remote server
+                // and 1 second of actual uploading.
+
+                QTimer::singleShot(UPLOAD_DELAY, [channel, &textEdit]() {
+                    uploadImageToNuuls(uploadQueue.front(), channel, textEdit);
+                    uploadQueue.pop();
+                });
+            }
+            return Success;
+        })
+        .onError([channel](NetworkResult result) -> bool {
+            channel->addMessage(makeSystemMessage(
+                QString("An error happened while uploading your image: %1")
+                    .arg(result.status())));
+            isUploading = false;
+            return true;
+        })
+        .execute();
+}
+
+void upload(const QMimeData *source, ChannelPtr channel,
+            ResizingTextEdit &outputTextEdit)
+{
+    // There's no need to thread proof this function. It is called only from the main thread.
+    if (isUploading)
+    {
+        channel->addMessage(makeSystemMessage(
+            QString("Please wait until the upload finishes.")));
+        return;
+    }
+
+    isUploading = true;
+    channel->addMessage(makeSystemMessage(QString("Started upload...")));
+
+    if (source->hasFormat("image/png"))
+    {
+        uploadImageToNuuls({source->data("image/png"), "png"}, channel,
+                           outputTextEdit);
+    }
+    else if (source->hasFormat("image/jpeg"))
+    {
+        uploadImageToNuuls({source->data("image/jpeg"), "jpeg"}, channel,
+                           outputTextEdit);
+    }
+    else if (source->hasFormat("image/gif"))
+    {
+        uploadImageToNuuls({source->data("image/gif"), "gif"}, channel,
+                           outputTextEdit);
+    }
+    else if (source->hasUrls())
+    {
+        // This path gets chosen when files are copied from a file manager, like explorer.exe, caja.
+        // Each entry in source->urls() is a QUrl pointing to a file that was copied.
+        for (const QUrl &path : source->urls())
+        {
+            QString localPath = path.toLocalFile();
+            if (!getImageFileFormat(localPath).isEmpty())
```

Comment:
We should replace the getImageFileFormat function entirely with Qt's [QMimeDatabase](https://doc.qt.io/qt-5/qmimedatabase.html) functions